plugins/cros-ec/fu-cros-ec-usb-device.c

```diff
@@ -578,6 +735,42 @@ fu_cros_ec_usb_device_prepare_firmware (FuDevice *device,
 	return g_steal_pointer (&firmware);
 }
 
+static gboolean
+fu_cros_ec_usb_device_attach (FuDevice *self, GError **error)
+{
+	g_debug ("cros_ec_usb_device_attach");
+
+	fu_device_set_remove_delay (self, FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	if (!fu_cros_ec_usb_device_jump_to_rw (self, error)) {
+
+		return FALSE;
+	}
+
+	fu_device_set_status (self, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_add_flag (self, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_detach (FuDevice *self, GError **error)
+{
+	g_debug ("cros_ec_usb_device_detach");
+
+	fu_device_set_remove_delay (self, FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	if (!fu_cros_ec_usb_device_reset_to_ro (self, error)) {
+
+		return FALSE;
+	}
+
+	fu_device_set_status (self, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_add_flag (self, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	/* success */
+	return TRUE;
+}
+
```

Comment:
Since these functions are 99% the same, just a thought - what if you collapse them into one that is used for both vfunc calls and use `self->writeable_offset` to tell which to call?

Reply:
Ah, since this was last uploaded, there's been more divergence between attach and detach paths.

Detach now checks for custom flags and bails out early if the rw region was written, but ro has not yet been written (bailing out will remain in RW). This is so that the RO write phase can happen.

Similarly, the attach phase has a slightly  different bailout criteria as well.